src/hotspot/share/prims/scopedMemoryAccess.cpp

```diff
@@ -138,17 +159,41 @@ class CloseScopedMemoryClosure : public HandshakeClosure {
 };
 
 /*
- * This function issues a global handshake operation with all
- * Java threads. This is useful for implementing asymmetric
- * dekker synchronization schemes, where expensive synchronization
- * in performance sensitive common paths, may be shifted to
- * a less common slow path instead.
- * Top frames containg obj will be deoptimized.
+ * This functin performs a thread-local handshake against all threads running at the time
+ * the given scope (deopt) was closed. If the handshake closure finds that a thread has
+ * safepointed inside a scoped method (that is, a method inside the ScopedMemoryAccess class
+ * annotated with the '@Scoped' annotation), whose local variables mention the scope being closed
+ * (deopt), the thread is added to a problematic stack. After the handshake, each thread in
+ * the problematic stack is handshaked again, individually, to check that it has exited
+ * the scoped method. This should happen quickly, because once we find a problematic
+ * thread, we also deoptimize it, meaning that when the thread resumes execution, it should
+ * also see the updated scope state. This function returns when the stack of problematic
+ * threads is empty. To prevent premature thread termination we take a snapshot of the live
+ * threads in the system using a ThreadsListHandle.
  */
-JVM_ENTRY(jboolean, ScopedMemoryAccess_closeScope(JNIEnv *env, jobject receiver, jobject deopt))
-  CloseScopedMemoryClosure cl(deopt);
+JVM_ENTRY(void, ScopedMemoryAccess_closeScope(JNIEnv *env, jobject receiver, jobject deopt))
+  ThreadStack threads;
+  CloseScopedMemoryClosure cl(deopt, &threads);
+  // do a first handshake and collect all problematic threads
   Handshake::execute(&cl);
-  return !cl._found;
+  if (threads.empty()) {
+    // fast-path: return if no problematic thread is found
+    return;
+  }
+  // now iterate on all problematic threads, until we converge
+  ThreadsListHandle tlh;
+  ThreadStackElement *element = threads.pop();
+  while (element != NULL) {
+    JavaThread* thread = element->_thread;
+    // if the thread is not in the list handle, we can safely skip further handshakes,
+    // as that means that the thread has been created when the scope state has already
+    // been set to CLOSED - meaning that the thread access will fail anyway.
```

Comment:
I don't get this comment...

It seems true that 
1. there might be threads in the thread list which are not in the problematic stack.
2. or vice versa, there can be threads in the stack that are not in the thread list.

---

1. seems to be covered by the fact that we are iterating the threads in the thread stack, instead of the ones in the list.
2. It seems that the `includes` check is needed since a thread might have died since the initial handshake, in which case we don't want to handshake it again?

So, I think the comment should read something regarding 2. instead. e.g. 
```suggestion
    // if the thread is not in the list handle, it means that it has died since the initial handshake.
    // in that case we don't need to handshake it again.
```

Reply:
The logic here has changed to reflect some internal comments - I will update the comment accordingly